File: ios/AmazonChimeSdkComponentModule.swift
import ExpoModulesCore
import AmazonChimeSDK
import AmazonChimeSDKMedia

public class AmazonChimeSdkComponentModule: Module {
    private var logger = ConsoleLogger(name: "AmazonChimeSdkComponent", level: .DEBUG)
    private var isAudioSessionStopped = true
    private var isVideoSessionStopped = true
    private var meetingSession: DefaultMeetingSession?

    private enum EventKeys {
        static let meetingStart = "OnMeetingStart"
        static let meetingEnd = "OnMeetingEnd"
        static let error = "OnError"
        static let success = "OnSuccess"
        static let cameraSwitchOn = "OnCameraSwitchOn"
        static let addVideoTile = "OnAddVideoTile"
        static let removeVideoTile = "OnRemoveVideoTile"
        static let attendeesJoin = "OnAttendeesJoin"
        static let attendeesLeave = "OnAttendeesLeave"
        static let attendeesMute = "OnAttendeesMute"
        static let attendeesUnmute = "OnAttendeesUnmute"
        static let dataMessageReceive = "OnDataMessageReceive"
        static let audioSessionStartConnecting = "OnAudioSessionStartConnecting"
    }

    public func definition() -> ModuleDefinition {
        Name("AmazonChimeSdkComponent")

        Events(
            EventKeys.meetingStart,
            EventKeys.meetingEnd,
            EventKeys.error,
            EventKeys.success,
            EventKeys.cameraSwitchOn,
            EventKeys.addVideoTile,
            EventKeys.removeVideoTile,
            EventKeys.attendeesJoin,
            EventKeys.attendeesLeave,
            EventKeys.attendeesMute,
            EventKeys.attendeesUnmute,
            EventKeys.dataMessageReceive,
            EventKeys.audioSessionStartConnecting
        )

        AsyncFunction("startMeeting") { (meetingInfo: [String: Any], attendeeInfo: [String: Any]) in
            self.stopMeetingIfExists()
            guard let sessionConfig = self.createSessionConfiguration(meetingInfo: meetingInfo, attendeeInfo: attendeeInfo) else {
                self.sendEvent(EventKeys.error, ["message": "Failed to create meeting session"])
                return
            }
            // let context = UIApplication.shared
            self.meetingSession = DefaultMeetingSession(configuration: sessionConfig, logger: self.logger)
            self.startAudioVideo()
            self.sendEvent(EventKeys.success, ["meetingId": "Ekpono ambrose"])
        }

        AsyncFunction("stopMeeting") {
            self.logger.info(msg: "Called stopMeeting")
            self.sendEvent(EventKeys.meetingEnd, ["meetingStopped": "Meeting has stopped"])
            self.meetingSession?.audioVideo.stop()

            self.isAudioSessionStopped = true
            self.isVideoSessionStopped = true

            // remove tile
            self.sendEvent(EventKeys.removeVideoTile, ["tileId": 0])
            self.meetingSession = nil
        }

        AsyncFunction("setMute") { (isMute: Bool) in
            self.logger.info(msg: "Called setMute: \(isMute)")
            if isMute {
                _ = self.meetingSession?.audioVideo.realtimeLocalMute()
            } else {
                _ = self.meetingSession?.audioVideo.realtimeLocalUnmute()
            }
        }

        AsyncFunction("setCameraOn") { (enabled: Bool) in
            self.sendEvent(EventKeys.cameraSwitchOn, ["camera": enabled])

            do {
                if enabled {
                    try _ = self.meetingSession?.audioVideo.startLocalVideo()
                } else {
                    try _ = self.meetingSession?.audioVideo.stopLocalVideo()
                }
            } catch {
                self.sendEvent(EventKeys.error, ["message": "Failed to start local video"])
            }
            

        }

        //switch camera
        AsyncFunction("switchCamera") {
            self.logger.info(msg: "Called switchCamera")
            self.meetingSession?.audioVideo.switchCamera()
        }


        View(AmazonChimeSdkComponentView.self) {
            Prop("tileId") { (view: AmazonChimeSdkComponentView, tileId: Int) in
                guard let session = self.meetingSession else {
                    self.logger.error(msg: "MeetingSession is null; cannot bind video view")
                    return
                }
                view.setTileId(tileId: tileId, meetingSession: session)
            }
        }
    }

    private func startVideo() {
        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .notDetermined:
            AVCaptureDevice.requestAccess(for: .video) { granted in
                if granted {
                    self.startVideo()
                } else {
                    self.sendEvent("OnError", ["message": "User denied camera permission"])
                }
            }
        case .authorized:
            do {
                try self.meetingSession?.audioVideo.startLocalVideo()
            } catch {
                self.sendEvent("OnError", ["message": "Failed to start local video"])
            }
        case .denied:
            UIApplication.shared.open(URL(string: UIApplication.openSettingsURLString)!)
        default:
            break
        }
    }

    private func stopMeetingIfExists() {
        if let session = meetingSession {
            session.audioVideo.stop()
            meetingSession = nil
        }
    }

    private func startAudioVideo() {
    do {
        guard let session = meetingSession else { return }
        session.audioVideo.addRealtimeObserver(observer: observer)
        session.audioVideo.addVideoTileObserver(observer: observer)
        session.audioVideo.addAudioVideoObserver(observer: observer)
        try session.audioVideo.start()
        _ = session.audioVideo.startRemoteVideo()
        sendEvent(EventKeys.meetingStart, ["meetingStarted": "Meeting started"])
        } catch {
            logger.error(msg: "Failed to start audioVideo session: \(error)")
            sendEvent(EventKeys.error, ["message": "Failed to start audio video"])
        }
    }


    let urlRewriter: URLRewriter = { url in
        return url // Just return the input URL without any changes
    }

    private func createSessionConfiguration(meetingInfo: [String: Any], attendeeInfo: [String: Any]) -> MeetingSessionConfiguration? {
        guard let meetingId = meetingInfo["MeetingId"] as? String,
              let attendeeId = attendeeInfo["AttendeeId"] as? String,
              let joinToken = attendeeInfo["JoinToken"] as? String,
              let externalUserId = attendeeInfo["ExternalUserId"] as? String,
              let mediaPlacement = meetingInfo["MediaPlacement"] as? [String: String],
              let audioFallbackUrl = mediaPlacement["AudioFallbackUrl"],
              let audioHostUrl = mediaPlacement["AudioHostUrl"],
              let turnControlUrl = mediaPlacement["TurnControlUrl"],
              let signalingUrl = mediaPlacement["SignalingUrl"] else {
            // logger.error(msg: "Missing or invalid session configuration data")
            return nil
        }

        let credentials = MeetingSessionCredentials(attendeeId: attendeeId, externalUserId: externalUserId, joinToken: joinToken)
        let urls = MeetingSessionURLs(audioFallbackUrl: audioFallbackUrl, audioHostUrl: audioHostUrl, turnControlUrl: turnControlUrl, signalingUrl: signalingUrl, urlRewriter: urlRewriter)
        return MeetingSessionConfiguration(meetingId: meetingId, credentials: credentials, urls: urls, urlRewriter: urlRewriter)
    }

    private lazy var observer = Observer(self)

    @objc class Observer: NSObject, RealtimeObserver, AudioVideoObserver, VideoTileObserver {
        private let module: AmazonChimeSdkComponentModule

        init(_ module: AmazonChimeSdkComponentModule) {
            self.module = module
        }

        func attendeesDidJoin(attendeeInfo: [AttendeeInfo]) {
            attendeeInfo.forEach { attendee in
                module.sendEvent(EventKeys.attendeesJoin, ["attendeeId": attendee.attendeeId, "externalUserId": attendee.externalUserId])
            }
        }

        func attendeesDidLeave(attendeeInfo: [AttendeeInfo]) {
            attendeeInfo.forEach { attendee in
                module.sendEvent(EventKeys.attendeesLeave, ["attendeeId": attendee.attendeeId, "externalUserId": attendee.externalUserId])
            }
        }

        func attendeesDidDrop(attendeeInfo: [AttendeeInfo]) {
            attendeeInfo.forEach { attendee in
                module.sendEvent(EventKeys.attendeesLeave, ["attendeeId": attendee.attendeeId, "externalUserId": attendee.externalUserId])
            }
        }

        func attendeesDidMute(attendeeInfo: [AttendeeInfo]) {
            attendeeInfo.forEach { attendee in
                module.sendEvent(EventKeys.attendeesMute, ["attendeeId": attendee.attendeeId])
            }
        }

        func attendeesDidUnmute(attendeeInfo: [AttendeeInfo]) {
            attendeeInfo.forEach { attendee in
                module.sendEvent(EventKeys.attendeesUnmute, ["attendeeId": attendee.attendeeId])
            }
        }

        func signalStrengthDidChange(signalUpdates: [SignalUpdate]) {
            signalUpdates.forEach { update in
                module.logger.info(msg: "Attendee \(update.attendeeInfo.attendeeId) signalStrength changed to \(update.signalStrength)")
            }
        }

        func volumeDidChange(volumeUpdates: [VolumeUpdate]) {
            volumeUpdates.forEach { update in
                module.logger.info(msg: "Attendee \(update.attendeeInfo.attendeeId) volumeLevel changed to \(update.volumeLevel)")
            }
        }

        func audioSessionDidStartConnecting(reconnecting: Bool) {
            module.logger.info(msg: "Audio session is starting. Reconnecting: \(reconnecting)")
            module.sendEvent(EventKeys.audioSessionStartConnecting, ["reconnecting": reconnecting])
        }

        func audioSessionDidStart(reconnecting: Bool) {
            module.logger.info(msg: "Audio session started. Reconnecting: \(reconnecting)")
            module.sendEvent(EventKeys.meetingStart, ["reconnecting": reconnecting])
        }

        func audioSessionDidDrop() {
            module.logger.info(msg: "Audio session dropped due to poor network conditions.")
            // module.sendEvent(EventKeys.audioSessionDropped, ["reason": "poor network conditions"])
        }

        func audioSessionDidStopWithStatus(sessionStatus: MeetingSessionStatus) {
            module.logger.info(msg: "Audio session stopped. Status: \(sessionStatus.statusCode)")
            module.sendEvent(EventKeys.meetingEnd, ["status": sessionStatus.statusCode.rawValue])
        }

        func audioSessionDidCancelReconnect() {
            module.logger.info(msg: "Audio reconnection was canceled.")
            // module.sendEvent(EventKeys.audioSessionCancelReconnect, ["reason": "canceled"])
        }

        func connectionDidRecover() {
            module.logger.info(msg: "Connection health recovered.")
            // module.sendEvent(EventKeys.connectionRecovered, ["status": "recovered"])
        }

        func videoSessionDidStartConnecting() {
            module.logger.info(msg: "Video session is connecting.")
            // module.sendEvent(EventKeys.videoSessionStartConnecting, ["status": "connecting"])
        }

        func videoSessionDidStartWithStatus(sessionStatus: MeetingSessionStatus) {
            module.logger.info(msg: "Video session started with status: \(sessionStatus.statusCode)")
            // module.sendEvent(EventKeys.videoSessionStart, ["status": sessionStatus.statusCode.rawValue])
        }

        func videoSessionDidStopWithStatus(sessionStatus: MeetingSessionStatus) {
            module.logger.info(msg: "Video session stopped with status: \(sessionStatus.statusCode)")
            // module.sendEvent(EventKeys.videoSessionStop, ["status": sessionStatus.statusCode.rawValue])
        }

        func remoteVideoSourcesDidBecomeAvailable(sources: [RemoteVideoSource]) {
            module.logger.info(msg: "Remote video sources became available.")
            // let sourceIds = sources.map { $0.sourceId }
            // module.sendEvent(EventKeys.remoteVideoSourcesAvailable, ["sources": sourceIds])
        }

        func remoteVideoSourcesDidBecomeUnavailable(sources: [RemoteVideoSource]) {
            module.logger.info(msg: "Remote video sources became unavailable.")
            // let sourceIds = sources.map { $0.sourceId }
            // module.sendEvent(EventKeys.remoteVideoSourcesUnavailable, ["sources": sourceIds])
        }

        func cameraSendAvailabilityDidChange(available: Bool) {
            module.logger.info(msg: "Camera send availability changed. Available: \(available)")
            // module.sendEvent(EventKeys.cameraSendAvailabilityChanged, ["available": available])
        }

        func connectionDidBecomePoor() {
            module.logger.info(msg: "Connection has become poor.")
            // module.sendEvent(EventKeys.connectionPoor, ["status": "poor"])
        }

        func videoTileDidAdd(tileState: VideoTileState) {
            module.logger.info(msg: "Video tile added with state: \(tileState)")
            // Send event with tile details
            module.sendEvent(EventKeys.addVideoTile, 
            [
                "tileId": tileState.tileId, 
                "attendeeId": tileState.attendeeId,
                "isLocal": tileState.isLocalTile,
                "isScreenShare": tileState.isContent,
                "pauseState": tileState.pauseState.rawValue,
                "videoStreamContentWidth": tileState.videoStreamContentWidth,
                "videoStreamContentHeight": tileState.videoStreamContentHeight
            ])
        }

        func videoTileDidRemove(tileState: VideoTileState) {
            module.logger.info(msg: "Video tile removed with state: \(tileState)")
            // Send event with tile ID
            module.sendEvent(EventKeys.removeVideoTile, ["tileId": tileState.tileId])
        }

        func videoTileDidPause(tileState: VideoTileState) {
            module.logger.info(msg: "Video tile paused with state: \(tileState)")
            // Send event to indicate tile pause
            // module.sendEvent(EventKeys.videoTileDidPause, ["tileId": tileState.tileId])
        }

        func videoTileDidResume(tileState: VideoTileState) {
            module.logger.info(msg: "Video tile resumed with state: \(tileState)")
            // Send event to indicate tile resume
            // module.sendEvent(EventKeys.videoTileDidResume, ["tileId": tileState.tileId])
        }

        func videoTileSizeDidChange(tileState: VideoTileState) {
            module.logger.info(msg: "Video tile size changed with state: \(tileState)")
            // Send event with new size details
            // module.sendEvent(EventKeys.videoTileSizeDidChange, ["tileId": tileState.tileId, "width": tileState.videoStreamContentWidth, "height": tileState.videoStreamContentHeight])
        }

    }

    private func emitMeetingEndEventIfNeeded() {
        if isAudioSessionStopped && isVideoSessionStopped {
            sendEvent(EventKeys.meetingEnd, ["meetingEnd": true])
        }
    }
}

File: android/src/main/java/expo/modules/amazonchimesdkcomponent/AmazonChimeSdkComponentModule.kt
package expo.modules.amazonchimesdkcomponent

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.AppContext
import com.amazonaws.services.chime.sdk.meetings.session.DefaultMeetingSession
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSession
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionConfiguration
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionCredentials
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionURLs
import com.amazonaws.services.chime.sdk.meetings.session.defaultUrlRewriter
import com.amazonaws.services.chime.sdk.meetings.audiovideo.AttendeeInfo
import com.amazonaws.services.chime.sdk.meetings.audiovideo.AudioVideoObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.SignalUpdate
import com.amazonaws.services.chime.sdk.meetings.audiovideo.VolumeUpdate
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.RemoteVideoSource
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileState
import com.amazonaws.services.chime.sdk.meetings.realtime.RealtimeObserver
import com.amazonaws.services.chime.sdk.meetings.realtime.datamessage.DataMessage
import com.amazonaws.services.chime.sdk.meetings.realtime.datamessage.DataMessageObserver
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionStatus
import com.amazonaws.services.chime.sdk.meetings.utils.logger.ConsoleLogger
import com.amazonaws.services.chime.sdk.meetings.utils.logger.LogLevel
import android.util.Log


class AmazonChimeSdkComponentModule : Module() {
    private val logger = ConsoleLogger(LogLevel.DEBUG)

    private var isAudioSessionStopped = true
    private var isVideoSessionStopped = true

    companion object {
        // Event types
        const val CHIME_EVENT_ERROR = "OnError"
        const val CHIME_EVENT_MEETING_START = "OnMeetingStart"
        const val CHIME_EVENT_MEETING_END = "OnMeetingEnd"
        const val CHIME_EVENT_VIDEO_TILE_ADD = "OnAddVideoTile"
        const val CHIME_EVENT_VIDEO_TILE_REMOVE = "OnRemoveVideoTile"
        const val CHIME_EVENT_ATTENDEES_JOIN = "OnAttendeesJoin"
        const val CHIME_EVENT_ATTENDEES_LEAVE = "OnAttendeesLeave"
        const val CHIME_EVENT_ATTENDEES_MUTE = "OnAttendeesMute"
        const val CHIME_EVENT_ATTENDEES_UNMUTE = "OnAttendeesUnmute"
        const val CHIME_EVENT_DATA_MESSAGE_RECEIVE = "OnDataMessageReceive"

        // Additional data for attendee events
        private const val CHIME_EVENT_KEY_ATTENDEE_ID = "attendeeId"
        private const val CHIME_EVENT_KEY_EXTERNAL_USER_ID = "externalUserId"

        // Additional data for video tile events
        private const val CHIME_EVENT_KEY_VIDEO_TILE_ID = "tileId"
        private const val CHIME_EVENT_KEY_VIDEO_IS_LOCAL = "isLocal"
        private const val CHIME_EVENT_KEY_VIDEO_IS_SCREEN_SHARE = "isScreenShare"
        private const val CHIME_EVENT_KEY_VIDEO_ATTENDEE_ID = "attendeeId"
        private const val CHIME_EVENT_KEY_VIDEO_PAUSE_STATE = "pauseState"
        private const val CHIME_EVENT_KEY_VIDEO_VIDEO_STREAM_CONTENT_HEIGHT = "videoStreamContentHeight"
        private const val CHIME_EVENT_KEY_VIDEO_VIDEO_STREAM_CONTENT_WIDTH = "videoStreamContentWidth"

        // Additional data for data message
        private const val CHIME_EVENT_KEY_DATA_MESSAGE_DATA = "data"
        private const val CHIME_EVENT_KEY_DATA_MESSAGE_SENDER_ATTENDEE_ID = "senderAttendeeId"
        private const val CHIME_EVENT_KEY_DATA_MESSAGE_SENDER_EXTERNAL_USER_ID = "senderExternalUserId"
        private const val CHIME_EVENT_KEY_DATA_MESSAGE_THROTTLED = "throttled"
        private const val CHIME_EVENT_KEY_DATA_MESSAGE_TIMESTAMP_MS = "timestampMs"
        private const val CHIME_EVENT_KEY_DATA_MESSAGE_TOPIC = "topic"

        private const val TAG = "ExpoMeetingObservers"

        private const val WEBRTC_PERMISSION_REQUEST_CODE = 1
        private const val KEY_MEETING_ID = "MeetingId"
        private const val KEY_ATTENDEE_ID = "AttendeeId"
        private const val KEY_JOIN_TOKEN = "JoinToken"
        private const val KEY_EXTERNAL_ID = "ExternalUserId"
        private const val KEY_MEDIA_PLACEMENT = "MediaPlacement"
        private const val KEY_AUDIO_FALLBACK_URL = "AudioFallbackUrl"
        private const val KEY_AUDIO_HOST_URL = "AudioHostUrl"
        private const val KEY_TURN_CONTROL_URL = "TurnControlUrl"
        private const val KEY_SIGNALING_URL = "SignalingUrl"
        private const val TOPIC_CHAT = "chat"

        var meetingSession: MeetingSession? = null

        private var videoView: AmazonChimeSdkComponentView? = null
    }

    override fun definition() = ModuleDefinition {
        Name("AmazonChimeSdkComponent")

        Events("OnMeetingStart", "OnMeetingEnd", "OnError", "OnAttendeesUnmute", "OnSuccess", "OnCameraSwitchOn",
        // Attendee event keys
        CHIME_EVENT_ERROR,
        CHIME_EVENT_MEETING_START,
        CHIME_EVENT_MEETING_END,
        CHIME_EVENT_VIDEO_TILE_ADD,
        CHIME_EVENT_VIDEO_TILE_REMOVE,
        CHIME_EVENT_ATTENDEES_JOIN,
        CHIME_EVENT_ATTENDEES_LEAVE,
        CHIME_EVENT_ATTENDEES_MUTE,
        CHIME_EVENT_ATTENDEES_UNMUTE,
        CHIME_EVENT_DATA_MESSAGE_RECEIVE
    )
        // Define functions for JavaScript to call
        AsyncFunction("startMeeting") { meetingInfo: Map<String, Any?>, attendeeInfo: Map<String, Any?> ->
            if (meetingSession != null) {
                meetingSession?.audioVideo?.stop()
                meetingSession = null
            }


            val sessionConfig = createSessionConfiguration(meetingInfo, attendeeInfo)

            // Initialize meetingSession correctly
            meetingSession = sessionConfig?.let { config ->
                appContext.reactContext?.let { context -> 
                    DefaultMeetingSession(config, logger, context)
                }
            }

            if (meetingSession != null) {
                startAudioVideo()
                sendEvent("OnSuccess", mapOf("meetingId" to meetingSession?.configuration?.meetingId))
            } else {
                sendEvent("OnError", mapOf("message" to "Failed to create meeting session"))
            }
        }

        AsyncFunction("stopMeeting") {
            logger.info("AmazonChimeSdkComponentModule", "Called stopMeeting")
            sendEvent("OnMeetingEnd", mapOf("meetingStopped" to "Meeting has stopped"))
            meetingSession?.audioVideo?.stop()
        }

        AsyncFunction("setMute") { isMute: Boolean ->
            if (isMute) {
                meetingSession?.audioVideo?.realtimeLocalMute()
            } else {
                meetingSession?.audioVideo?.realtimeLocalUnmute()
            }
        }

        AsyncFunction("setCameraOn") { enabled: Boolean ->
            logger.info("AmazonChimeSdkComponentModule", "Called setCameraOn: $enabled")
            sendEvent("OnCameraSwitchOn", mapOf("camera" to enabled))
            if (enabled) {
                meetingSession?.audioVideo?.startLocalVideo()
            } else {
                meetingSession?.audioVideo?.stopLocalVideo()
            }
        }

        AsyncFunction("sendDataMessage") { topic: String, message: String, lifetimeMs: Int ->
            meetingSession?.audioVideo?.realtimeSendDataMessage(topic, message, lifetimeMs)
        }

        // swich camera
        AsyncFunction("switchCamera") {
            meetingSession?.audioVideo?.switchCamera()
        }

        View(AmazonChimeSdkComponentView::class) {
            Prop("tileId") { view: AmazonChimeSdkComponentView, tileId: Int ->
                meetingSession?.let { session ->
                    view.setTileId(tileId, session)
                } ?: logger.error("AmazonChimeSdkComponentModule", "MeetingSession is null; cannot bind video view")
            }

            // Prop("isOnTop") { view: AmazonChimeSdkComponentView, isOnTop: Boolean ->
            //     view.setZOrderMediaOverlay(isOnTop)
            // }
        }
    }

    // Function to send attendee update events
    public fun sendAttendeeUpdateEvent(eventName: String, attendeeInfo: AttendeeInfo) {
        val eventData = mapOf(
            CHIME_EVENT_KEY_ATTENDEE_ID to attendeeInfo.attendeeId,
            CHIME_EVENT_KEY_EXTERNAL_USER_ID to attendeeInfo.externalUserId
        )
        sendEvent(eventName, eventData)
    }

    private val observer = object : RealtimeObserver, VideoTileObserver, AudioVideoObserver, DataMessageObserver {

        override fun onAttendeesJoined(attendeeInfo: Array<AttendeeInfo>) {
            Log.d(TAG, "onAttendeesJoined called with ${attendeeInfo.size} attendees")
            
            attendeeInfo.forEach { attendee ->
                Log.i(TAG, "Processing join for attendee: ${attendee.attendeeId}")

                val eventData = mapOf(
                    "attendeeId" to attendee.attendeeId,
                    "externalUserId" to attendee.externalUserId,
                )

                sendEvent(CHIME_EVENT_ATTENDEES_JOIN, eventData)
            }
        }
        
      
    
        override fun onAttendeesLeft(attendeeInfo: Array<AttendeeInfo>) {
            attendeeInfo.forEach { attendee ->
                Log.i(TAG, "Processing join for attendee: ${attendee.attendeeId}")

                val eventData = mapOf(
                    "attendeeId" to attendee.attendeeId,
                    "externalUserId" to attendee.externalUserId
                )
                sendEvent(CHIME_EVENT_ATTENDEES_LEAVE, eventData)
            }
        }
    
        override fun onAttendeesDropped(attendeeInfo: Array<AttendeeInfo>) {
            attendeeInfo.forEach { attendee ->
                Log.i(TAG, "Processing join for attendee: ${attendee.attendeeId}")

                val eventData = mapOf(
                    "attendeeId" to attendee.attendeeId,
                    "externalUserId" to attendee.externalUserId
                )

                sendEvent(CHIME_EVENT_ATTENDEES_LEAVE, eventData)
            }
        }
    
        override fun onAttendeesMuted(attendeeInfo: Array<AttendeeInfo>) {
            attendeeInfo.forEach { attendee ->
                Log.i(TAG, "Processing join for attendee: ${attendee.attendeeId}")

                val eventData = mapOf(
                    "attendeeId" to attendee.attendeeId,
                    "externalUserId" to attendee.externalUserId
                )

                sendEvent(CHIME_EVENT_ATTENDEES_MUTE, eventData)
            }
        }
    
        override fun onAttendeesUnmuted(attendeeInfo: Array<AttendeeInfo>) {
            attendeeInfo.forEach { attendee ->
                Log.i(TAG, "Processing join for attendee: ${attendee.attendeeId}")

                val eventData = mapOf(
                    "attendeeId" to attendee.attendeeId,
                    "externalUserId" to attendee.externalUserId
                )

                sendEvent(CHIME_EVENT_ATTENDEES_UNMUTE, eventData)
            }
        }
    
        override fun onSignalStrengthChanged(signalUpdates: Array<SignalUpdate>) {
            // Not implemented for demo purposes
        }
    
        override fun onVolumeChanged(volumeUpdates: Array<VolumeUpdate>) {
            // Not implemented for demo purposes
        }
    
        override fun onVideoTileAdded(tileState: VideoTileState) {
            meetingSession?.let { session ->
                videoView?.setTileId(tileState.tileId, session)
            } ?: logger.error(TAG, "meetingSession or videoView is null; cannot set tileId")
        
            sendEvent(CHIME_EVENT_VIDEO_TILE_ADD, videoTileStateToMap(tileState))
        }
    
        override fun onVideoTilePaused(tileState: VideoTileState) {
            // Not implemented for demo purposes
        }
    
        override fun onVideoTileRemoved(tileState: VideoTileState) {
            sendEvent(CHIME_EVENT_VIDEO_TILE_REMOVE, videoTileStateToMap(tileState))
        }
    
        override fun onVideoTileResumed(tileState: VideoTileState) {
            // Not implemented for demo purposes
        }
    
        override fun onVideoTileSizeChanged(tileState: VideoTileState) {
            // Not implemented for demo purposes
        }
    
        override fun onAudioSessionCancelledReconnect() {
            // Not implemented for demo purposes
        }
    
        override fun onAudioSessionDropped() {
            // Not implemented for demo purposes
        }
    
        override fun onAudioSessionStarted(reconnecting: Boolean) {
            logger.info(TAG, "Received event for audio session started. Reconnecting: $reconnecting")
    
            if (!reconnecting) {
                isAudioSessionStopped = false
                isVideoSessionStopped = false
                sendEvent(CHIME_EVENT_MEETING_START, mapOf("meetingStop" to reconnecting))
            }
        }
    
        override fun onAudioSessionStartedConnecting(reconnecting: Boolean) {
            // Not implemented for demo purposes
        }
    
        override fun onAudioSessionStopped(sessionStatus: MeetingSessionStatus) {
            // Not implemented for demo purposes
            isAudioSessionStopped = true
            emitRnMeetingEndEventIfNeeded()
        }
    
        override fun onCameraSendAvailabilityUpdated(available: Boolean) {
            // Not implemented for demo purposes
        }
    
        override fun onConnectionBecamePoor() {
            // Not implemented for demo purposes
        }
    
        override fun onConnectionRecovered() {
            // Not implemented for demo purposes
        }
    
        override fun onVideoSessionStarted(sessionStatus: MeetingSessionStatus) {
            // Not implemented for demo purposes
        }
    
        override fun onVideoSessionStartedConnecting() {
            // Not implemented for demo purposes
        }
    
        override fun onVideoSessionStopped(sessionStatus: MeetingSessionStatus) {
            // Not implemented for demo purposes
            isVideoSessionStopped = true
            emitRnMeetingEndEventIfNeeded()
        }
    
        override fun onDataMessageReceived(dataMessage: DataMessage) {
            // eventEmitter.sendDataMessageEvent(CHIME_EVENT_DATA_MESSAGE_RECEIVE, dataMessage)
        }
    
        override fun onRemoteVideoSourceAvailable(sources: List<RemoteVideoSource>) {
            // Not implemented for demo purposes
        }
    
        override fun onRemoteVideoSourceUnavailable(sources: List<RemoteVideoSource>) {
            // Not implemented for demo purposes
        }
    
        fun emitRnMeetingEndEventIfNeeded() {
            if(isAudioSessionStopped && isVideoSessionStopped) {
                sendEvent(CHIME_EVENT_MEETING_END, mapOf("meetingEnd" to true))
            }
        }
    }

    private fun startAudioVideo() {
        meetingSession?.let {
            it.audioVideo.addRealtimeObserver(observer) // this observer is crashing the app
            it.audioVideo.addVideoTileObserver(observer)
            it.audioVideo.addAudioVideoObserver(observer)
            it.audioVideo.start()
            it.audioVideo.startRemoteVideo()
            sendEvent("OnMeetingStart", mapOf("meeting-started" to "meeting started"))
        }
    }

    private fun createSessionConfiguration(meetingInfo: Map<String, Any?>, attendeeInfo: Map<String, Any?>): MeetingSessionConfiguration? {
        return try {
            val meetingId = meetingInfo["MeetingId"] as? String ?: ""
            val attendeeId = attendeeInfo["AttendeeId"] as? String ?: ""
            val joinToken = attendeeInfo["JoinToken"] as? String ?: ""
            val externalUserId = attendeeInfo["ExternalUserId"] as? String ?: ""

            val mediaPlacement = meetingInfo["MediaPlacement"] as Map<String, String>
            val audioFallbackUrl = mediaPlacement?.get("AudioFallbackUrl") ?: ""
            val audioHostUrl = mediaPlacement?.get("AudioHostUrl") ?: ""
            val turnControlUrl = mediaPlacement?.get("TurnControlUrl") ?: ""
            val signalingUrl = mediaPlacement?.get("SignalingUrl") ?: ""

            MeetingSessionConfiguration(
                meetingId,
                MeetingSessionCredentials(attendeeId, externalUserId, joinToken),
                MeetingSessionURLs(audioFallbackUrl, audioHostUrl, turnControlUrl, signalingUrl, ::defaultUrlRewriter)
            )
        } catch (exception: Exception) {
            logger.error("AmazonChimeSdkComponentModule", "Error creating session configuration: ${exception.localizedMessage}")
            sendEvent("OnError", mapOf("message" to "Error creating session configuration: ${exception.localizedMessage}"))
            null
        }
    }
    // Function to send data message events
    public fun sendDataMessageEvent(eventName: String, dataMessage: DataMessage) {
        val eventData = mapOf(
            CHIME_EVENT_KEY_DATA_MESSAGE_DATA to dataMessage.text(),
            CHIME_EVENT_KEY_DATA_MESSAGE_SENDER_ATTENDEE_ID to dataMessage.senderAttendeeId,
            CHIME_EVENT_KEY_DATA_MESSAGE_SENDER_EXTERNAL_USER_ID to dataMessage.senderExternalUserId,
            CHIME_EVENT_KEY_DATA_MESSAGE_THROTTLED to dataMessage.throttled,
            CHIME_EVENT_KEY_DATA_MESSAGE_TIMESTAMP_MS to dataMessage.timestampMs.toDouble(),
            CHIME_EVENT_KEY_DATA_MESSAGE_TOPIC to dataMessage.topic
        )
        sendEvent(eventName, eventData)
    }

    // Function to send video tile events
    public fun sendVideoTileEvent(eventName: String, tileState: VideoTileState) {
        val eventData = mapOf(
            CHIME_EVENT_KEY_VIDEO_TILE_ID to tileState.tileId,
            CHIME_EVENT_KEY_VIDEO_IS_LOCAL to tileState.isLocalTile,
            CHIME_EVENT_KEY_VIDEO_IS_SCREEN_SHARE to tileState.isContent,
            CHIME_EVENT_KEY_VIDEO_ATTENDEE_ID to tileState.attendeeId,
            CHIME_EVENT_KEY_VIDEO_PAUSE_STATE to tileState.pauseState.ordinal,
            CHIME_EVENT_KEY_VIDEO_VIDEO_STREAM_CONTENT_HEIGHT to tileState.videoStreamContentHeight,
            CHIME_EVENT_KEY_VIDEO_VIDEO_STREAM_CONTENT_WIDTH to tileState.videoStreamContentWidth
        )
        sendEvent(eventName, eventData)
    }

    private fun emitRnMeetingEndEventIfNeeded() {
        if(isAudioSessionStopped && isVideoSessionStopped) {
            sendEvent(CHIME_EVENT_MEETING_END, mapOf("meetingEnd" to true))
        }
    }

    private fun attendeeInfoToMap(attendeeInfo: AttendeeInfo): Map<String, Any?> {
        return mapOf(
            "attendeeId" to attendeeInfo.attendeeId,
            "externalUserId" to attendeeInfo.externalUserId
        )
    }
    
    // Conversion for VideoTileState
    private fun videoTileStateToMap(videoTileState: VideoTileState): Map<String, Any?> {
        return mapOf(
            "tileId" to videoTileState.tileId,
            "isLocal" to videoTileState.isLocalTile,
            "isScreenShare" to videoTileState.isContent,
            "attendeeId" to videoTileState.attendeeId,
            "pauseState" to videoTileState.pauseState.ordinal,
            "videoStreamContentHeight" to videoTileState.videoStreamContentHeight,
            "videoStreamContentWidth" to videoTileState.videoStreamContentWidth
        )
    }
}
